appcues/src/main/java/com/appcues/ui/ModalStateMachineOwner.kt

```diff
@@ -0,0 +1,167 @@
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import java.util.Timer
+import java.util.TimerTask
+import kotlin.concurrent.schedule
+
+internal class ModalStateMachineOwner(
+    override val stateMachine: StateMachine,
+    private val coroutineScope: AppcuesCoroutineScope,
+) : StateMachineOwning {
+
+    companion object {
+        private const val SCROLL_DEBOUNCE_MILLISECONDS = 1_000L
+    }
+
+    override val renderContext: RenderContext = Modal
+
+    // used to attempt retry when items in view change
+    private var viewTreeUpdateHandler: ViewTreeUpdateHandler = ViewTreeUpdateHandler { attemptRetry() }
+    private var uiIdleDebounceTimer: TimerTask? = null
+    private var viewTreeObserver: ViewTreeObserver? = null
+
+    init {
+        // set up observer on the Modal state machine errors to initiate retry
+        coroutineScope.launch(Dispatchers.IO) {
+            stateMachine.errorFlow.collect { error ->
+                // if a recoverable step error is observed, begin attempting retry
+                if (error is StepError && error.recoverable) {
+                    startRetryHandling()
+                }
+            }
+        }
+    }
+
+    override suspend fun reset() {
+        stateMachine.stop(true)
+    }
+
+    fun stopRetryHandling() {
+        viewTreeUpdateHandler.detach()
+        viewTreeObserver = null
+    }
+
+    private fun startRetryHandling() {
+        if (viewTreeObserver == null) {
+            AppcuesActivityMonitor.activity?.getParentView()?.viewTreeObserver?.let {
+                viewTreeUpdateHandler.attach(it)
+                this.viewTreeObserver = it
+            }
+        }
+    }
+
+    private fun attemptRetry() {
+        onUiIdle {
+            // cancel any current touch (drag) as our overlay is about to drop on top and we don't want things moving behind it
+            AppcuesActivityMonitor.activity?.getParentView()?.dispatchTouchEvent(
+                MotionEvent.obtain(0L, 0L, MotionEvent.ACTION_CANCEL, 0f, 0f, 0)
+            )
+            coroutineScope.launch {
+                // stop listening for updates, will re-attach if another error occurs
+                stopRetryHandling()
+                stateMachine.handleAction(Retry)
+            }
+        }
+    }
+
+    private fun onUiIdle(block: () -> Unit) {
+        // cancel any previous
+        uiIdleDebounceTimer?.cancel()
+
+        uiIdleDebounceTimer = Timer().schedule(delay = SCROLL_DEBOUNCE_MILLISECONDS) {
+            // set to know to allow new Timer
+            uiIdleDebounceTimer = null
+            // run callback block
+            block()
+        }
+    }
+
+    private class ViewTreeUpdateHandler(private var action: () -> Unit) :
+        OnScrollChangedListener,
+        OnGlobalLayoutListener,
+        OnTouchModeChangeListener,
+        OnDrawListener,
+        OnGlobalFocusChangeListener,
+        OnWindowAttachListener,
+        OnWindowFocusChangeListener
```

Comment:
not all of these are used (they were used for logging an experiementation) - this will be cleaned up on any final pass over this code to trim down to the set of listeners needed. At this point it appears that scroll change, draw, and global layout listeners will suffice - and watching those until they become Idle gives us an indicator that we can re-query the UI for missing elements